File: BOJ(Java 공부용)/18808.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0, 1, -1, 0};
    static int[] goy = {1, 0, 0, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken()), k = Integer.parseInt(st.nextToken());
        int[][] board = new int[n][m];
        List<int[][]> sticker = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());
            int r = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());
            int[][] temp = new int[r][c];
            for (int j = 0; j < r; j++) {
                st = new StringTokenizer(br.readLine());
                for (int p = 0; p < c; p++) {
                    temp[j][p] = Integer.parseInt(st.nextToken());
                }
            }
            sticker.add(temp);
        }

        for (int[][] e : sticker) {
            int r = e.length;
            int c = e[0].length;

            // 1번
            if (first(n, m, board, e, r, c)) {
                continue;
            }
            //2번(돌리기)
            for (int i = 0; i < 3; i++) {
                int[][] temp;
                if (i == 0) {
                    temp = new int[c][r];
                    for (int a = 0, q = 0; a < c; a++, q++) {
                        for (int b = 0, p = r - 1; b < r; b++, p--) {
                            temp[a][b] = e[p][q];
                        }
                    }
                    if (first(n, m, board, temp, c, r)) {
                        break;
                    }
                } else if (i == 1) {
                    temp = new int[r][c];
                    for (int a = 0, p = r - 1; a < r; a++, p--) {
                        for (int b = 0, q = c - 1; b < c; b++, q--) {
                            temp[a][b] = e[p][q];
                        }
                    }
                    if (first(n, m, board, temp, r, c)) {
                        break;
                    }
                } else {
                    temp = new int[c][r];
                    for (int a = 0, q = c - 1; a < c; a++, q--) {
                        for (int b = 0, p = 0; b < r; b++, p++) {
                            temp[a][b] = e[p][q];
                        }
                    }
                    if (first(n, m, board, temp, c, r)) {
                        break;
                    }
                }
            }
        }
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 1) cnt++;
            }
        }

        bw.write(cnt + "");
        bw.flush();
        bw.close();
    }

    private static boolean first(int n, int m, int[][] board, int[][] sticker, int r, int c) {
        int ex = n - r, ey = m - c;
        if (ex < 0 || ey < 0) return false;
        int x = 0, y = 0;
        boolean ok = false;
        while (true) {
            if (x > ex) break;
            int p = x;
            int q = y;
            boolean no = false;
            for (int i = 0, a = p; i < r; i++, a++) {
                for (int j = 0, b = q; j < c; j++, b++) {
                    if (board[a][b] == 1 && sticker[i][j] == 1) {
                        no = true;
                    }
                }
            }
            if (!no) { //맞는것이 존재
                for (int i = x, a = 0; i < x + r; i++, a++) {
                    for (int j = y, b = 0; j < y + c; j++, b++) {
                        if (sticker[a][b] == 1 && board[i][j] == 0)
                            board[i][j] = sticker[a][b];
                    }
                }
                ok = true;
                break;
            }
            if (y == ey) {
                y = 0;
                x++;
            } else y++;
        }

        return ok;
    }
}

File: BOJ(Python 공부용)/18808.py
import sys

input = sys.stdin
sys.setrecursionlimit(5000)

n, m, k = map(int, input.readline().split())
board = [[0] * m for _ in range(n)]
sticker = []


def first(sticker, r, c):
    ex = n - r
    ey = m - c
    if ex < 0 or ey < 0: return False
    x = 0
    y = 0
    ok = False
    while True:
        p = x
        q = y
        no = False
        for i, a in zip(range(r), range(p, p + r, 1)):
            for j, b in zip(range(c), range(q, q + c, 1)):
                if board[a][b] == 1 and sticker[i][j] == 1:
                    no = True
        if not no:
            for i, a in zip(range(x, x + r, 1), range(0, r, 1)):
                for j, b in zip(range(y, y + c, 1), range(0, c, 1)):
                    if sticker[a][b] == 1 and board[i][j] == 0:
                        board[i][j] = sticker[a][b]
            ok = True
        if y == ey:
            y = 0
            x += 1
        else:
            y += 1
        if x > ex or ok:
            break
    return ok


for _ in range(k):
    r, c = map(int, input.readline().split())
    temp = [[int(x) for x in input.readline().split()] for _ in range(r)]

    sticker.append(temp)

for e in sticker:
    r = len(e)
    c = len(e[0])
    # 1번
    if first(e, r, c):
        continue

    # 2번
    for i in range(3):
        temp = [[]]
        if i == 0:
            temp = [[0] * r for _ in range(c)]
            for a, q in zip(range(c), range(0, c, 1)):
                for b, p in zip(range(r), range(r - 1, -1, -1)):
                    temp[a][b] = e[p][q]

            if first(temp, c, r):
                break
        if i == 1:
            temp = [[0] * c for _ in range(r)]
            for a, p in zip(range(r), range(r - 1, -1, -1)):
                for b, q in zip(range(c), range(c - 1, -1, -1)):
                    temp[a][b] = e[p][q]

            if first(temp, r, c):
                break
        if i == 2:
            temp = [[0] * r for _ in range(c)]
            for a, q in zip(range(c), range(c - 1, -1, -1)):
                for b, p in zip(range(r), range(0, r, 1)):
                    temp[a][b] = e[p][q]
            if first(temp, c, r):
                break

cnt = 0
for i in range(n):
    for j in range(m):
        if board[i][j] == 1:
            cnt += 1

print(cnt)